pdns/dnsdistdist/doh.cc

```diff
@@ -0,0 +1,718 @@
+  else {
+    return d_path.compare(0, d_path.size(), dq->du->req->path.base, dq->du->req->query_at) == 0;
+  }
+}
+
+string HTTPPathRule::toString() const
+{
+  return "url path == " + d_path;
+}
+
+
+void dnsdistclient(int qsock, int rsock)
+{
+  for(;;) {
+    try {
+      DOHUnit* du;
+      recv(qsock, &du, sizeof(du), 0);
+      // if there was no EDNS, we add it with a large buffer size
+      // so we can use UDP to talk to the backend.
+      struct dnsheader* dh = (struct dnsheader*)du->query.c_str();
+
+      if(!dh->arcount) {
+        const uint8_t name = 0;
+        dnsrecordheader drh;
+        EDNS0Record edns0;
+        edns0.extRCode = 0;
+        edns0.version = 0;
+        edns0.extFlags = 0;
+        
+        drh.d_type = htons(QType::OPT);
+        drh.d_class = htons(4096);
+        memcpy(&drh.d_ttl, &edns0, sizeof edns0);
+        drh.d_clen = htons(0);
+        string res;
+        res.assign((const char *) &name, sizeof name);
+        res.append((const char *) &drh, sizeof drh);
+        
+        du->query += res;
+        dh = (struct dnsheader*)du->query.c_str(); // may have reallocated
+        dh->arcount = htons(1);
+        
+        du->ednsAdded = true;
+      }
+      else {
+        // we leave existing EDNS in place
+      }
+      
+      if(processDOHQuery(du) < 0) {
+        du->error = true; // turns our drop into a 500
+        if(send(du->rsock, &du, sizeof(du), 0) != sizeof(du))
+          delete du;     // XXX but now what - will h2o time this out for us?
+      }
+    }
+    catch(std::exception& e) {
+      errlog("Error while processing query received over DoH: %s", e.what());
+    }
+    catch(...) {
+      errlog("Unspecified error while processing query received over DoH");
+    }
+  }
+}
+
+// called if h2o finds that dnsdist gave us an answer
+static void on_dnsdist(h2o_socket_t *listener, const char *err)
+{
+  DOHUnit *du;
+  DOHServerConfig* dsc = (DOHServerConfig*)listener->data;
+  recv(dsc->dohresponsepair[1], &du, sizeof(du), 0);
+
+  if(!du->req) { // it got killed in flight
+//    cout << "du "<<(void*)du<<" came back from dnsdist, but it was killed"<<endl;
+    delete du;
+    return;
+  }
+  *du->self = 0; // so we don't clean up again in on_generator_dispose
+  if(!du->error) {
+    dsc->df->d_validresponses++;
+    du->req->res.status = 200;
+    du->req->res.reason = "OK";
+
+    h2o_add_header(&du->req->pool, &du->req->res.headers, H2O_TOKEN_CONTENT_TYPE, NULL, H2O_STRLIT("application/dns-message"));
+    
+    //    struct dnsheader* dh = (struct dnsheader*)du->query.c_str();
+    //    cout<<"Attempt to send out "<<du->query.size()<<" bytes over https, TC="<<dh->tc<<", RCODE="<<dh->rcode<<", qtype="<<du->qtype<<", req="<<(void*)du->req<<endl;
+
+    if(du->ednsAdded) { // take it out again
+      vector<uint8_t> stripped;      
+      rewriteResponseWithoutEDNS(du->query, stripped);
+      du->query.assign((const char*)&stripped[0], stripped.size());
+    }
```

Comment:
The case of an incoming query with `EDNS` but without `ECS` is not handled here.